libminifi/include/io/ZlibStream.h

```diff
@@ -0,0 +1,105 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LIBMINIFI_INCLUDE_IO_ZLIBSTREAM_H_
+#define LIBMINIFI_INCLUDE_IO_ZLIBSTREAM_H_
+#include <cstdint>
+#include <vector>
+
+#include <zlib.h>
+
+#include "BaseStream.h"
+#include "core/logging/LoggerConfiguration.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace io {
+
+enum class ZlibStreamState : uint8_t {
+  UNINITIALIZED,
+  INITIALIZED,
+  ERRORED,
+  FINISHED
+};
+
+class ZlibBaseStream : public BaseStream {
+ public:
+  ZlibBaseStream();
+  ZlibBaseStream(DataStream* other);
+
+  ZlibBaseStream(const ZlibBaseStream&) = delete;
+  ZlibBaseStream& operator=(const ZlibBaseStream&) = delete;
+  ZlibBaseStream(ZlibBaseStream&& other) = delete;
+  ZlibBaseStream& operator=(ZlibBaseStream&& other) = delete;
+
+  ~ZlibBaseStream() override = default;
+
+  virtual bool isFinished() const;
+
+ protected:
+  ZlibStreamState state_{ZlibStreamState::UNINITIALIZED};
+  z_stream strm_{};
+  std::vector<uint8_t> outputBuffer_;
+};
+
+class ZlibCompressStream : public ZlibBaseStream {
+ public:
+  ZlibCompressStream(bool gzip = true, int level = Z_DEFAULT_COMPRESSION);
+  ZlibCompressStream(DataStream* other, bool gzip = true, int level = Z_DEFAULT_COMPRESSION);
```

Comment:
Fair point. I'll replace the `bool gzip` with an enum class to the effect of `ZlibFormat format`. A pointer is not converted implicitly to an enum class, and it will also improve readability.

Reply:
Could you also mark both ctors as `explicit`? Since this is new code, it doesn't break API compatibility.

Reply:
I could, although I don't really see the point in this case.

Reply:
To prevent implicit conversion from `bool`/`ZlibFormat` and `DataStream*`, as suggested in [C.46: By default, declare single-argument constructors explicit](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit). (The guideline doesn't, but the reason covers all constructors that can be called with a single argument, i.e. including ctors with more parameters where all N>1-th parameter has a default argument.)

Or if you intended the implicit conversion to be part of the API, I'm also fine with an explanation and possibly an example of intended usage added in a code comment.

Reply:
Fair, will mark them explicit.